docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+This document is a draft of changes to ECMA-335 for the "runtime async" feature. When the feature is officially supported, it can be merged into the final ECMA-335 augments document.
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async methods have a special signature encoding, described in [### I.8.6.1.5 Method signatures].
+
+Sync methods are all other methods.
+
+Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
+
+Async methods support the following suspension points:
+
+* Calling another async method. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
+* Using a library in the .NET corelib to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
+  ```C#
+  // public static async2 Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  // public static async2 Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  ```
+
+Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, in can be presumed that the task has completed.
+
+Async methods have the following restrictions:
+* Usage of the `localloc` and instruction is forbidden
+* The `lodloca` and `ldarga` instructions are redefined to return managed pointers instead of pointers.
+* Pinning locals may not be created
+* The `tail` prefix is forbidden
```

Comment:
https://github.com/dotnet/csharplang/discussions/8031#discussioncomment-9000799

F# requires tail call support. Is there any plan to supported it afterwards, if not in initial release?

Reply:
Yes, this will be crucial for us, and we also want to support it on release.

Reply:
I think some further work is required from my side before it is possible to answer this question, in particular around what kind of constraints the compatibility handling is going to put on the semantics of async2 methods.
For example, if compatibility ends up requiring us to have a side-effecting epilog in all async2 methods then that makes tailcall support hard/impossible.

If we ignore potential compatibility concerns and use a model close to what exists in the current prototype, then I do not see a significant barrier to allowing `tail.` (famous last words). We may just need an async2 aware version of the tailcall dispatcher that knows about the potential of either suspension or a follow-up call.

It might be that F# is just going to opt out of all the compatibility scaffolding, in which case maybe this will be simple work (and the final documentation of the feature should just document how `tail.` interacts with the different compatibility switches).

@vzarytovskii if you have existing concrete `async` cases where forms of tailcalls are required for correctness it would be very helpful for me to look at how we translate those into the async2 model.

Reply:
It is expected that there are a number of existing C# scenarios that will not be possible in `runtime async` and the old codegen will need to be preserved for those cases. "Async iterators" is one such example. I think `tail.` is another. The F# compiler would be expected to use it's old codegen for cases that are not initially supported.

We may expand the scope later, but that's not part of this document.

Reply:
I was wondering about async-iterators too. They could still take advantage of async2, but that would require new codegen from roslyn (ie. generate an `async2 MoveNextAsync`)